File: expo-screen-time/src/ExpoScreenTimeModule.ts
import { NativeModule, requireNativeModule } from 'expo';

import { ScreenTimeEvents } from './ExpoScreenTime.types';
import { AppUsageStats } from './ExpoScreenTime.types';

declare class ExpoScreenTimeModule extends NativeModule<ScreenTimeEvents> {
  checkPermission(): boolean;
  requestPermission(): void;
  getUsageForApps(packageNames: string[], sinceTimestamp: number): Promise<AppUsageStats>;
}

// Returns null in Expo Go (no native module) instead of throwing.
let module: ExpoScreenTimeModule | null = null;
try {
  module = requireNativeModule<ExpoScreenTimeModule>('ExpoScreenTime');
} catch {
  // Native module not available (e.g. running in Expo Go)
}

export default module;

File: expo-screen-time/ios/ExpoScreenTimeModule.swift
import ExpoModulesCore

public class ExpoScreenTimeModule: Module {
  // TODO: iOS Screen Time API requires Apple's Family Controls entitlement
  // This entitlement needs special approval from Apple and is only available
  // for apps that meet specific criteria (parental control, health apps, etc.)
  //
  // To implement real screen time tracking on iOS:
  // 1. Request Family Controls entitlement from Apple
  // 2. Add the entitlement to your Apple Developer account
  // 3. Use ScreenTime API (DeviceActivity and FamilyActivityPicker)
  // 4. Follow Apple's guidelines for parental control features
  //
  // Reference: https://developer.apple.com/documentation/screen_time
  //
  // For now, this module returns mock data to demonstrate the API structure.

  public func definition() -> ModuleDefinition {
    Name("ExpoScreenTime")

    // Check if screen time permission is granted
    // On iOS, this would check Family Controls authorization
    Function("checkPermission") {
      return false // TODO: Return real permission status after implementing Family Controls
    }

    // Open iOS Settings to request permission
    // On iOS, this would open Screen Time settings
    Function("requestPermission") {
      // TODO: Implement Family Controls authorization flow
      // For now, we can't open Screen Time settings programmatically
    }

    // Get usage stats for specified apps since a given timestamp
    // Returns mock data for demonstration
    AsyncFunction("getUsageForApps") { (packageNames: [String], sinceTimestamp: Double) in
      // TODO: Implement real screen time tracking using DeviceActivity framework
      // Reference: https://developer.apple.com/documentation/deviceactivity

      // Return mock data for now
      var mockStats: [String: Double] = [:]

      for packageName in packageNames {
        // Generate mock usage time between 0 and 3600 seconds (1 hour)
        // In a real implementation, this would query actual usage data
        mockStats[packageName] = Double.random(in: 0...3600)
      }

      return mockStats
    }
  }
}
